pynumaflow/sourcer/async_server.py

```diff
@@ -0,0 +1,213 @@
+import logging
+import multiprocessing
+import os
+
+from collections.abc import AsyncIterable
+from google.protobuf import timestamp_pb2 as _timestamp_pb2
+import grpc
+from google.protobuf import empty_pb2 as _empty_pb2
+
+from pynumaflow import setup_logging
+from pynumaflow._constants import (
+    MAX_MESSAGE_SIZE,
+    SOURCE_SOCK_PATH,
+)
+from pynumaflow.sourcer import Datum
+from pynumaflow.sourcer._dtypes import AsyncSourceReadCallable, Offset, AckRequest
+from pynumaflow.sourcer.proto import source_pb2
+from pynumaflow.sourcer.proto import source_pb2_grpc
+from pynumaflow.types import NumaflowServicerContext
+from pynumaflow.info.server import get_sdk_version, write as info_server_write
+from pynumaflow.info.types import ServerInfo, Protocol, Language, SERVER_INFO_FILE_PATH
+
+_LOGGER = setup_logging(__name__)
+if os.getenv("PYTHONDEBUG"):
+    _LOGGER.setLevel(logging.DEBUG)
+
+_PROCESS_COUNT = multiprocessing.cpu_count()
+MAX_THREADS = int(os.getenv("MAX_THREADS", 0)) or (_PROCESS_COUNT * 4)
```

Comment:
4 is a magic number. Please add comments on why we chose 4 instead of other numbers.